### ISSUE TITLE
Apply MRS spectral leak correction

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#7786]
_Issue [JP-3327](https://jira.stsci.edu/browse/JP-3327) was created on JIRA by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw):_

As has long been known, the MRS has a spectral leak in which 6 micron light leaks into the 12 micron channel.  A correction for this has long been part of the pipeline design and spec3 workflow (see <https://outerspace.stsci.edu/display/JWSTCC/Vanilla+Spectral+Leak+Subtraction),> but it is only now with the recent overhaul of MRS flux calibration that such a correction has become possible.

(Previously the leak was baked into the photom calibration reference files, while now it is not, so a correction relies only upon the spectrum of the science target and not the calibration source).

The algorithm for this already exists and has been tested during the flux calibration process and independently by some science teams.  In brief, a relative throughput reference file would have to be added to CRDS, multiplied by the Ch1B spectrum, and subtracted from the Ch3A spectrum.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=657434#comment-657434):_

Work investigating this has indicated that the placeholder reference mrsptcorr reference file is appropriate to use for this correction.  However, it should have its data model updated in order to better support this correction (both for short-term notebooks and long-term pipeline implementation).

Added ticket <https://jira.stsci.edu/browse/CRDS-721> for any CRDS work necessary, and submitted spacetelescope/stdatamodels#196 for the data model change.  As noted in the CRDS ticket, this change should have no impact on pipeline or operations as it is a placeholder reference file not currently used in any stage of the pipeline.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=660342#comment-660342):_

What's the status of this? Is the datamodel update for the ref file needed for the DMS B10.0 delivery (coming up in 2 weeks)?

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=660355#comment-660355):_

Looks like the next thing needed is a review of spacetelescope/stdatamodels#196

I've updated this to reflect recent changes to the master branch.  If we can get this in for B10.0 that would be great, but not critical as this doesn't affect the pipeline.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=660723#comment-660723):_

Datamodels update complete: spacetelescope/stdatamodels#196

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=664496#comment-664496):_

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) Could you point me to the algorithm for this ticket ?

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=664499#comment-664499):_

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) Also do you have some good data to test this with. I am thinking something with a large leak from 6 to 12

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=664509#comment-664509):_

See <https://github.com/STScI-MIRI/MRS-ExampleNB/blob/main/D2P_Notebooks/SpectralLeak.ipynb> for implementation

Any standard star observation should work well for testing, e.g. PID 1538 Obs 1.

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=664516#comment-664516):_

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) I have questions on how this should be implemented in the pipeline.  I am assuming this is in calspec 3 at the end. In general the pipeline does not make  band specific s3d files and it seems the algorithm needs ch1b x1d created from ch1b s3d and also a ch3a s3d & x1d.

So what is intension of how this should work in calspec3. Can we adapt the correction to use ch1 x1d and apply it ch3 x1d - since those are default products from spec3  or are you thinking something else ?

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=664833#comment-664833):_

It should work fine with any spectra covering the 1B and 3A wavelength ranges; the other bands are just irrelevant to the calculation.  So, it should be fine to apply it to a 1ABC and 3ABC spectrum for instance.  The idea is that this would happen after Extract1D.  Extract1D is often called on a set of cubes covering multiple wavelengths, if so then the leak correction would use whatever spectrum covered the 1B range to compute a correction for anything covering the 3A range.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=664834#comment-664834):_

I.e., whether the output is per-band, per-channel, or all-wavelength it should still be possible to make the correction.  Should probably only be applied for POINT sources by default though.

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=666709#comment-666709):_

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw)  I am going to make a new step to apply the spectral leak. By default it will be run in calspec3 on MRS data. This step will take the output from extract1d . I am thinking if the results from extract1d are for a point source and  contain a spectrum with 1b and a separate one for  3A the step will be run- otherwise skip.
But maybe we also want a few options to the step - like providing the name of the 1B x1d file to use to correct the 3A x1d.
Do you have any preference on things like:
1. step name: spectral_leak
2. optional parameters to the step ?

3. Anything I might have forgotten.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=666726#comment-666726):_

That sounds like a good place to start.  Shouldn't be any optional parameters, as there are really no tunable parameters in the step.  Overrides possible for the reference file and the reference spectrum.

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=667804#comment-667804):_

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw)  I am sending the spectral leak step all the x1d files returned from an assocation in calwebb_spec3.
From this set of x1d files I need to set the data that covers 1B and the set that covers 3 A. 
I am researching this but I don't think I can just query the channel and band datamodel values because the input data is multuspec and those values are not part of the schema. In addition the band  might have a value of 'multi.
So I was thinking of just reading in the  wavelength array and hard coding a 1B wavelength = 6 microns

3A wavelength = 12. If the x1d covers either one of those values then I know which X1d to use. 
Sound ok ?

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=667811#comment-667811):_

As [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) mentioned in the previous comment, the x1d products coming out of the calwebb_spec3 extract_1d step do not have useful meta information to tell you which bands were included in creating the 1D spectrum. Because data from multiple bands have been combined, it just tells you that it comes from multiple bands, period, without saying which ones. For example, an x1d file for channel 1 data:
```
In [6]: x1d.meta.instrument.instance
Out[6]: 
{'band': 'MULTIPLE',
 'ccc_state': 'OPEN',
 'channel': '1',
 'detector': 'MIRIFUSHORT',
 'lamp_mode': 'NONE',
 'lamp_state': 'OFF',
 'miri_fpe_side': 'A',
 'miri_ice_side': 'A',
 'name': 'MIRI'}```

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=668122#comment-668122):_

[Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) Sounds reasonable.  Perhaps look for all input spectra covering the range 5.8-6.5 microns, and apply it to spectra in the range 10.6-13 microns.  That should cover the entire range we care about.  If multiple inputs in the relevant range, average them.  If multiple outputs, apply to all of them.

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=668435#comment-668435):_

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) 

Maybe this is fine. I can get the R_MRSPT reference file written and S_SPLEAK set to COMPLETE if when I run the calwebb_spec3 I set on the command line --steps.spectral_leak.save_results=True



If I don't save_results= True I don't get either in the header.
I think this means that we will need to update the jwst_miri_pars_spec3pipeline to now include 
spectral_leak.

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=668448#comment-668448):_

[Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse)  [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) I WAS WRONG you do not set save_results=True. I was getting confused because the results from extract1d are being written out and then new files from spectral leak correction are also being written out with  difference names. I don't want a second set written out just 1 final set. So I need to figure out what why this second set is being written out.

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=668451#comment-668451):_

[Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) 

Could you clarify how the spectral leak step should run. After extract_1d the extracted files from the s3d files are written to disk. If there exists the correct data to perform the spectral leak correction (which will only correct the spectra containing the CH 3A data) do we want to override the existing X1d for that CH or write a new file  and should the step only write the spectra containing the Ch3A data.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=668453#comment-668453):_

I defer to [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on what is best in keeping with the overall pipeline philosophy and easiest to implement, but to my mind it would either overwrite the relevant x1d files or simply not write out the x1d files at all until after the step was run (holding them in memory in the same way that other intermediary data models are when files are not written out).  I think having a new optional filetype would both represent unnecessary work for multiple groups (e.g., MAST and DMS) and also clutter things up from a user standpoint.  Much like aperture correction, it's either a step that's been applied to the 1d spectrum or not.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=668459#comment-668459):_

I tend to agree that the x1d files should be updated to contain the corrected results, rather than saving a whole new product type. It doesn't matter to me whether we let the extract_1d step continue to save its results, which then get overwritten by the leak correction step, or just not write out the x1d files till after leak corr has been performed. If we do the latter, we'd probably need some additional logic in calwebb_spec3 in between the extract_1d and leak corr steps for it to figure out when to write the x1d files. For example, if we're processing NRS IFU or NRS MOS data, the leak corr step doesn't apply and hence we need to save things after extract_1d. If the extract_1d step itself is doing the writing of the files, then I guess there's no choice but to have leak corr overwrite them (or at least some of them).

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=668468#comment-668468):_

 

Something is hidden someplace that is not letting me run extract_1d and not write out the output until after the spectral_leak correction is done. I think it is the save_results=True

in the 

jwst_miri_pars-spec3pipeline_0005.asdf


 I have tried to force the filenames to be be the same but something is overriding that and

giving the spectral leak correction results a different file name.

Here is the code in calwebb_spec3:
if exptype in IFU_EXPTYPES:
                    self.extract_1d.search_output_file = False
                    if exptype in ['MIR_MRS']:
                        result = self.extract_1d(result)
                        result = self.spectral_leak(result)
                    else:
                        result = self.extract_1d(result)

 

HOW DO I PREVENT THE X1D files from being written after extract_1d step ?

This the log file and you can see the file names are different (I have not idea where the file names are being formed for the spectral leak step - it is hidden).

2023-10-23 16:32:35,908 - stpipe.Spec3Pipeline.extract_1d - INFO - Saved model in jw01538-o001_t019_miri_test_ch4-short_x1d.fits

2023-10-23 16:32:35,957 - stpipe.Spec3Pipeline.extract_1d - INFO - Saved model in jw01538-o001_t019_miri_test_ch2-shortmedium_x1d.fits

2023-10-23 16:32:36,009 - stpipe.Spec3Pipeline.extract_1d - INFO - Saved model in jw01538-o001_t019_miri_test_ch1-shortmedium_x1d.fits

2023-10-23 16:32:36,062 - stpipe.Spec3Pipeline.extract_1d - INFO - Saved model in jw01538-o001_t019_miri_test_ch3-short_x1d.fits

2023-10-23 16:32:36,062 - stpipe.Spec3Pipeline.extract_1d - INFO - Step extract_1d done

2023-10-23 16:32:36,259 - stpipe.Spec3Pipeline.spectral_leak - INFO - Step spectral_leak running with args (<ModelContainer>,).

2023-10-23 16:32:36,260 - stpipe.Spec3Pipeline.spectral_leak - INFO - Step spectral_leak parameters are: \{'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': True, 'skip': False, 'suffix': 'x1d', 'search_output_file': True, 'input_dir': ''}

2023-10-23 16:32:36,270 - stpipe.Spec3Pipeline.spectral_leak - INFO - Input is a ModelContainer

2023-10-23 16:32:36,426 - stpipe.Spec3Pipeline.spectral_leak - INFO - Saved model in jw01538-o001_t019_miri_test_0_x1d.fits 

2023-10-23 16:32:36,474 - stpipe.Spec3Pipeline.spectral_leak - INFO - Saved model in jw01538-o001_t019_miri_test_1_x1d.fits

2023-10-23 16:32:36,523 - stpipe.Spec3Pipeline.spectral_leak - INFO - Saved model in jw01538-o001_t019_miri_test_2_x1d.fits

2023-10-23 16:32:36,571 - stpipe.Spec3Pipeline.spectral_leak - INFO - Saved model in jw01538-o001_t019_miri_test_3_x1d.fits

2023-10-23 16:32:36,571 - stpipe.Spec3Pipeline.spectral_leak - INFO - Step spectral_leak done

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=669251#comment-669251):_

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw)  I made a few tests that used different combinations of bands to test the spectral leak correction. I do get some slight differences in the correction depending which combinations of bands I use.  Here are the different combinations I use

1. IFUSHORT Band SHORTMEDIUM and IFULONG Band Short
2. IFUSHORT Band MEDIUM and IFULONG Band Short

3. IFUSHORT Band SHORTMEDIUMLONG and IFULONG Band Short
4. IFUSHORT Band SHORTMEDIUMLONG and IFULONG Band ShortMedium
5. IFUSHORT Band SHORTMEDIUMLONG and IFULONG Band ShortMediumLong

Test 1 & 2 give similar tests and Test 3-5 give similar results. There is a slight offset between the two different overall results.  I am not sure if the slight different is "good enough" or if I should add something to the how the correction is determined or applied
I hope these plots are self explanatory: (top plot has all the results to show the difference)

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=669357#comment-669357):_

Hm, I think this warrants further investigation as there shouldn't (in principle) be a difference.  I wonder if what we're seeing is issues to do with the astrometric shift between bands.  I.e., including the 1C wave channel messing up the centroiding for 1B slightly, affecting the extracted spectrum?  Is there a test branch I can look at?

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=669374#comment-669374):_

Current PR is at:

#8039 

 

You will also need PR for stdatamodels spacetelescope/stdatamodels#228

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=670639#comment-670639):_

[David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw)  the stdatamodels work has been added to master. What further testing can I do for this PR ?  When I do not apply the MRS spectral leak I do get all the extracted spectra to line up on top of one another using combinations of the different bands - so I don't think it is a centroiding issue.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=671425#comment-671425):_

Adding a note here for posterity on what the bug above ended up being.  In brief, the example code needed to multiply the Ch1B wavelengths by two since it's second-order light spilling into Ch3A.  If not doing so the interpolation is close enough (and the correction small enough) that it still looks good for short wavelength segment inputs but will fail for longer segments.

With this fix the code looks good.  leak_compare.png shows before/after correction for both BAND and CHANNEL inputs.

**Comment by stscijgbot-jp** [comment]
_Comment by [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=673517#comment-673517):_

[Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) 

I am trying to make the spectral leak correction run  standalone (strun spectral_leak blah.asn.json).



I can get it to run if I make a spectral_leak.cfg file. But if I just trying and run spectral_leak I get an error. 
I am missing something - do you have any clue ? 



Here is the error: 

(dev58) arizonajane4:JP-3327_Spectral_Leak morrison$ strun spectral_leak  test_spectral_leak_asn.json

----------------------------------------------------------------------

ERROR PARSING CONFIGURATION:

    Config file spectral_leak not found.

----------------------------------------------------------------------

usage: strun [--logcfg LOGCFG] [--verbose] [--debug] [--save-parameters SAVE_PARAMETERS] [--disable-crds-steppars] cfg_file_or_class

 

Run an stpipe Step or Pipeline

 

positional arguments:

  cfg_file_or_class     The configuration file or Python class to run

 

options:

  --logcfg LOGCFG       The logging configuration file to load

  --verbose, -v         Turn on all logging messages

  --debug               When an exception occurs, invoke the Python debugger, pdb

  --save-parameters SAVE_PARAMETERS

                        Save step parameters to specified file.

  --disable-crds-steppars

                        Disable retrieval of step parameter references files from CRDS

Traceback (most recent call last):

  File "/Users/morrison/miniconda3/envs/dev58/lib/python3.10/site-packages/stpipe/cli/strun.py", line 24, in main

    Step.from_cmdline(sys.argv[1:])

  File "/Users/morrison/miniconda3/envs/dev58/lib/python3.10/site-packages/stpipe/step.py", line 187, in from_cmdline

    return cmdline.step_from_cmdline(args)

  File "/Users/morrison/miniconda3/envs/dev58/lib/python3.10/site-packages/stpipe/cmdline.py", line 381, in step_from_cmdline

    step, step_class, positional, debug_on_exception = just_the_step_from_cmdline(

  File "/Users/morrison/miniconda3/envs/dev58/lib/python3.10/site-packages/stpipe/cmdline.py", line 238, in just_the_step_from_cmdline

    step_class, config, name, config_file = _get_config_and_class(

  File "/Users/morrison/miniconda3/envs/dev58/lib/python3.10/site-packages/stpipe/cmdline.py", line 41, in _get_config_and_class

    config = config_parser.load_config_file(config_file)

  File "/Users/morrison/miniconda3/envs/dev58/lib/python3.10/site-packages/stpipe/config_parser.py", line 107, in load_config_file

    raise ValueError(f"Config file \{config_file} not found.")

ValueError: Config file spectral_leak not found.

(dev58) arizonajane4:JP-3327_Spectral_Leak morrison$

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=675475#comment-675475):_

Fixed by #8039

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3327?focusedCommentId=677769#comment-677769):_

Tested and working.
